File: dog/command/purge.go
package command

import (
	"github.com/disgoorg/disgo/discord"
	"github.com/disgoorg/snowflake/v2"
	"github.com/unfoco/dog/cmd"
)

type Purge struct {
	Count uint `cmd:"count"`
}

func (p Purge) Allow(ctx *cmd.Context) bool {
	return true //ctx.HasPermUser(discord.PermissionManageChannels)
}

func (p Purge) Run(ctx *cmd.Context, b *discord.MessageCreateBuilder) {
	if p.Count > 100 {
		b.SetContent("you can't purge more than 100 message at once")
		return
	}

	ids, err := p.GetMessageIds(ctx, p.Count)
	if err != nil {
		b.SetContent("unable to retrieve messages")
		return
	}

	err = ctx.Client.Rest().BulkDeleteMessages(ctx.ChannelID, ids)
	if err != nil {
		b.SetContent("unable to purge messages")
	} else {
		b.SetContentf("purged %v messages", len(ids))
	}
}

func (Purge) GetMessageIds(ctx *cmd.Context, count uint) ([]snowflake.ID, error) {
	list, err := ctx.Client.Rest().GetMessages(
		ctx.ChannelID,
		snowflake.ID(0),
		ctx.MessageID,
		snowflake.ID(0),
		int(count)+1,
	)
	if err != nil {
		return nil, err
	}

	var ids []snowflake.ID
	for _, v := range list {
		ids = append(ids, v.ID)
	}
	return ids, nil
}

File: src/command/misc/purge.rs
use crate::types;

#[poise::command(prefix_command, slash_command, required_permissions = "MANAGE_MESSAGES")]
pub async fn purge(
    ctx: types::Context<'_>,
    #[max = 50] count: u64,
) -> Result<(), types::Error> {

    if count > 50 {
        ctx.reply("you can't purge more than 50 messages at once").await?;
        return Ok(())
    }

    let messages = ctx
        .channel_id()
        .messages(ctx, |g| {
            g.limit(count)
        })
        .await?;

    ctx
        .channel_id()
        .delete_messages(ctx, messages)
        .await?;

    ctx.reply(format!("removed {} messages", count)).await?;
    Ok(())
}
